### ISSUE TITLE
Option to resize the overlay.

### ISSUE THREAD

**Issue by domdfcoding** [godot-extended-libraries/godot-debug-menu#27]
Is there (or could there be added) an option to resize or scale the overlay? My game uses a small viewport (400x300 in an 800x600 window) but the overlay fills the whole screen (and hangs off the left hand side if not the bottom. This is what it looks like:

![image](https://github.com/godot-extended-libraries/godot-debug-menu/assets/8050853/3b352d37-a12c-46ca-a9df-4795ebfae756)

If the overlay could be scaled down it would be much more useful.

**Comment by Calinou** [comment]
There isn't such an option yet, but it should be straightforward to multiply all font sizes according to a project setting.

Note that when using the `viewport` stretch mode, readability would suffer unless you used a font that is specifically designed to be readable at low sizes.

**Comment by LeeWannacott** [comment]
@Calinou I don't understand how to read the debug graph. Like for example in screenshot the Total is `16.42` for the average, but his CPU is `1.29` +` 3.82` for GPU, which would equal `5.11` which leaves like `~11mspf` unaccounted for? It doesn't make any sense to me,  Or am I reading this wrong?

**Comment by Calinou** [comment]
> Note that I have the opposite problem -- it's too small! 😅 Definitely need a resize feature. I'm running on a 5k display on mac.

Use the `canvas_items` stretch mode in the project settings, as explained in the [Multiple resolutions](https://docs.godotengine.org/en/stable/tutorials/rendering/multiple_resolutions.html) documentation. This will also handle the rest of your UI scaling automatically.

**Comment by smedelyan** [comment]
I'm using `canvas_items` stretch mode + 4k viewport which renders into 1080p window. The size is way too small as well.

**Comment by Calinou** [comment]
> I'm using `canvas_items` stretch mode + 4k viewport which renders into 1080p window. The size is way too small as well.

You can increase the Stretch Scale project setting since Godot 4.2, although this will affect all 2D elements (including the ones that are unrelated to the debug menu).

**Comment by smedelyan** [comment]
> although this will affect all 2D elements

Yep, I would like to avoid this :-) In fact, at the moment DebugMenu is the only thing in the project that is not scaled up properly.

**Comment by radiKal07** [comment]
Same issue, all UI is correct except the Debug Menu is too small and it's hard to read

**Comment by Calinou** [comment]
If you change **Default Theme Scale** in the Project Settings, it should increase the size of the debug menu text at least.

**Comment by antzGames** [comment]
I have a version of debug menu that lets the `F4` key cycle through 4 levels of CanvasLayer scale (0.75, 1, 1.5, 2).

The code is ready in a branch in my fork at: https://github.com/antzGames/godot-debug-menu/tree/Scaling-overlay-support

Open to suggestions, feedback @Calinou and all.

Some ideas to make it better are configurable scaling enum sets (but always have the default scale of 1 included).

**Comment by Calinou** [comment]
This should be done by changing the font size of all Labels and changing the graphs' size in code instead. Control or CanvasLayer scaling should **not** be used here. It's more work, but it ensures everything remains crisp.

**Comment by antzGames** [comment]
I agree for font crispness, this is the way to go, but it is going to be a lot code.

In addition to the font sizes and graph sizes:

- the hard coded font_outline_size also has to be scaled.
- the header widths for the FrameTimeHistory GridContainer node has to be scaled.
- hard coded Graph labels custom_minimum height of 27, which spaces the graphs vertically, has to be scaled.
- the graph size is set in a `const` and is used everywhere.  Do you still want 150 data points in each graph?  If we increase the size of the graph GUI by 2x, then the resolution of the polyline will suffer.

**Comment by antzGames** [comment]
✅OK, its complete.

So I used Groups for labels, header labels, and graph UI elements, and it was not that much code to implement.

**Features:**
✅ Users can set any font size with `DebugMenu.set_font_size()`.  The font size range is [12-72] . Default is 12pt.
✅ Font size, font outline size, header widths, and graphs are scaled correctly based on font size.
✅ F4 key cycles through 12pt (default), 18pt, 24pt, 30pt, 36pt font sizes and the CUSTOM font size that was set with `DebugMenu.set_font_size()`.  This gives 1x, 1.5x, 2x, 2.5x, 3x scaling without the user needing to use code.  This toggle only cycles with F4 if the DebugMenu is visible.

**Limitations:**
- So the font does not look good at any font size smaller than the default, which is 12pt.  My solution is for users who want a bigger overlay.

Code at: https://github.com/antzGames/godot-debug-menu/tree/Scaling-overlay-support

This is a test that sets font size with `DebugMenu.set_font_size()` between the range of 12 - 48, every 0.25s.  Look ok?

https://github.com/user-attachments/assets/a2b7adb6-e94d-40b2-bc17-1acb85549141

**Comment by Calinou** [comment]
> This is a test that sets font size with DebugMenu.set_font_size() between the range of 12 - 48, every 0.25s. Look ok?

Looks great 🙂 

I think there should also be a project setting to change the size at startup. To me, this is more important than having a key to change the size at runtime, as this will allow using the debug menu in projects with a low fixed viewport resolution (e.g. pixel art style). I think you can keep the runtime toggle key still, as it's useful for projects that use the `disabled` stretch mode.

I expect most pixel art projects to be using a size between 0.25 and 0.5, with an appropriate pixel art font that is readable at small sizes.

**Comment by antzGames** [comment]
> I think there should also be a project setting to change the size at startup.

✅Done!✅

You can now set the font size from the Project Settings.  No runtime code needed.  Good suggestion!

**Comment by antzGames** [comment]
> I expect most pixel art projects to be using a size between 0.25 and 0.5, with an appropriate pixel art font that is readable at small sizes.

- I added 0.25 (3pt) and 0.5 (6pt) in the `Display_Size` enum 
- I am testing with `stretch_mode = canvas_items`
- testing with a 400 x 300 viewport
- I set `font_outline_size` to 0 for these modes and it looks ok

**New issues:** ✅ all fixed ✅ 
- the default theme has `theme_override_constants/line_spacing` at 3, and for these small scales they are too large with too much white space between lines in `Information` and `Settings` labels. ✅ FIXED✅ 
- The main DebugMenu `Control` node position is also too large (-16, 8) and also needs to shrink for low scales.✅ No change needed✅ This padding is needed on mobile screens, rounded corners

**Comment by Calinou** [comment]
Out of curiosity, why does the initial font size need to be an enum rather than an integer? Many pixel fonts expect an exact size to be used (e.g. 8 or 10), which isn't available in the enum here.

**Comment by antzGames** [comment]
I removed the enum list and just settled for a INT project setting called custom_font_size.

The initial value is 12, with a range of 3-72.

BTW, how to I comment ProjectSettings properties so that the user gets more info when hovering over the property?

<img width="1372" height="1297" alt="Image" src="https://github.com/user-attachments/assets/42f84768-7865-4bec-9f75-022dd0f86a1f" />

**Comment by Calinou** [comment]
> BTW, how to I comment ProjectSettings properties so that the user gets more info when hovering over the property?

This isn't supported yet: https://github.com/godotengine/godot-proposals/discussions/8224

But here, the setting is obvious enough that a description isn't really needed.

I would rename the setting to just "Font Size" though, since the default value is intended to be used if you leave it to the default.

**Comment by antzGames** [comment]
> I would rename the setting to just "Font Size" though, since the default value is intended to be used if you leave it to the default.

✅Done✅ See update picture above.

So that brings up the next steps:

- would be great if you can do a code review to see if you want any other fields/functions renamed.
- who changes the `CHANGELOG.md`?  Sorry, I did not read the contributor guidelines.
- the README.md might need to be updated, who does that?

**Comment by Calinou** [comment]
> - would be great if you can do a code review to see if you want any other fields/functions renamed.

I suggest opening a PR. This way, I can leave review comments directly there.

> - who changes the CHANGELOG.md? Sorry, I did not read the contributor guidelines.
> - the README.md might need to be updated, who does that?

I'll handle these two when publishing a new release, so you don't have to do it.

**Comment by antzGames** [comment]
Implemented on my fork's master: https://github.com/antzGames/Antz-Debug-Menu

https://github.com/user-attachments/assets/ad3fce70-f1b9-476b-a569-f5e2044fc7eb
